### PR TITLE
Reset backoff duration after successful connection.

### LINKED ISSUE #554: Reset backoff interval after successful connection
Since #184, we retry long-poll failures, with exponential backoff up to an interval of 10 seconds.

Even after the device is again able to reach the server, though, the backoff interval on next failure remains 10 seconds, for as long as the app keeps running. That seems unhelpful. Instead, when we find we can again consistently reach the server, the backoff intervals should reset to what they would be if the app were restarted.

The simplest thing to do here would be: as soon as there's one more success, just reset the backoff to its initial state. I'm not sure if a more complex strategy would be best in general. But I think for these long-poll requests the simple thing is fine, for the reasons below.

## Detailed reasoning

The possible risk in completely resetting backoff on a single success is that if the server is having trouble and handling some but not all its requests, that plan would mean we'd do a lot more retries, partway toward what would happen if we weren't backing off at all. I'm not sure if there's a standard recommendation for how to strike a balance there.

However, these particular requests are a long-poll — which means that a *successful* request most often takes a long time to come back: up to 60 seconds. Probably most of the time it's longer than our max backoff interval of 10 seconds. So if after that, our next request randomly fails (because the server is having trouble) and we retry in 100ms, that's not that different from if we retry after 10 seconds.

### PR DESCRIPTION
**Pull request by PIG208**
This fixes #554, and supersedes #564.

### CODE REVIEW

File: test/api/backoff_test.dart [hunk 1]
```diff
@@ -14,27 +14,28 @@ Future<Duration> measureWait(Future<void> future) async {
...
+void verifyRandomizedDurations({
+  required List<int> expectedMaxDurationsInMs,
+  required List<Duration> Function(List<Duration>) computeActualDurations,
+}) {
     // This is a randomized test.  [numTrials] is chosen so that the failure
-    // probability < 1e-9.  There are 2 * 11 assertions, and each one has a
-    // failure probability < 1e-12; see below.
+    // probability < 1e-9 for up to 1000 assertions.
+    // There are 2 * [expectedMaxDurationsInMs.length] assertions,
+    // and each one has a failure probability < 1e-12; see below.
```
Comment (chrisbobbe) [inline test/api/backoff_test.dart:29]: The part "for up to 1000 assertions" is new in this commit. What does it mean, and is there a logic change it corresponds to?

**Review by chrisbobbe — COMMENTED** [review]
Thanks! One comment below, and I'll also mark this for Greg's review.

File: test/api/backoff_test.dart (hunk 1 above)
Comment (PIG208) [reply test/api/backoff_test.dart:29]: This commit allows an arbitrary number of expected durations to be passed, so the `1000` is here to convey how many expected values can be passed for the upper bound of 1e-9 to still hold.

**Comment by PIG208** [comment]
After having a discussion with Greg, we will drop the testing part of this because this change is easy to verify, but testing is complicated.

**Comment by gnprice** [comment]
Thanks @PIG208, and thanks @chrisbobbe for the previous review!

Merging, with an additional commit adding a long comment:
cca6f244a store [nfc]: Explain why we reset backoff, and the trade-offs

**Pull request merged**